.github/workflows/buildAndTest.yml

```diff
@@ -57,12 +57,63 @@ jobs:
     # Installing the results into the cache is an action which is automatically
     # added by the cache action above.
 
+  # --- end of build-llvm-standalone job.
+
+  build-llvm:
+    name: Build LLVM
+    runs-on: ubuntu-latest
+    steps:
+      # Clone the CIRCT repo and its submodules. Do shallow clone to save clone
+      # time.
+      - name: Get CIRCT
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 2
+          submodules: "true"
+
+      # Extract the LLVM submodule hash for use in the cache key.
+      - name: Get LLVM Hash
+        id: get-llvm-hash
+        run: echo "::set-output name=hash::$(git rev-parse @:./llvm)"
+        shell: bash
+
+      # Try to fetch LLVM from the cache.
+      - name: Cache LLVM
+        id: cache-llvm
+        uses: actions/cache@v2
+        with:
+          path: llvm
+          key: ${{ runner.os }}-llvm-prebuild-${{ steps.get-llvm-hash.outputs.hash }}
+
+      # Build LLVM if we didn't hit in the cache.
+      - name: Rebuild and Install LLVM
+        if: steps.cache-llvm.outputs.cache-hit != 'true'
+        run: |
+          mkdir build
+          cd build
+          cmake ../llvm/llvm \
+            -DLLVM_BUILD_EXAMPLES=OFF \
+            -DLLVM_TARGETS_TO_BUILD="host" \
+            -DLLVM_ENABLE_PROJECTS='mlir' \
+            -DLLVM_OPTIMIZED_TABLEGEN=ON \
+            -DLLVM_ENABLE_OCAMLDOC=OFF \
+            -DLLVM_ENABLE_BINDINGS=OFF \
+            -DCMAKE_C_COMPILER=clang \
+            -DCMAKE_CXX_COMPILER=clang++ \
+            -DLLVM_ENABLE_LLD=ON \
+            -DCMAKE_BUILD_TYPE=Release \
+            -DLLVM_ENABLE_ASSERTIONS=ON
+          cmake --build . -- -j$(nproc)
+
```

Comment:
It looks like you're just rebuilding LLVM with the same options, but a different cache.  This is going to have several impacts: increasing the time it takes to update LLVM, and also may hit the 5GB cache limit.  It's unclear to me from reading the doc whether a cache entry can be updated after it is created, so this may be unnecessary.  At the very least, it seems like we can create a second cache entry starting from the first cache by using restore-keys:.  I think we need a better solution here.

Reply:
Is it really necessary to test both build strategies in the PR gate? I would rather keep it as is since it works well with caching.